be/src/exec/workgroup/scan_task_queue.cpp

```diff
@@ -47,168 +41,169 @@ void ScanTaskQueueWithWorkGroup::close() {
     _cv.notify_all();
 }
 
-void ScanTaskQueueWithWorkGroup::_cal_wg_cpu_real_use_ratio() {
-    int64_t total_run_time = 0;
-    std::vector<int64_t> growth_times;
-    growth_times.reserve(_ready_wgs.size());
-    for (auto& wg : _ready_wgs) {
-        growth_times.emplace_back(wg->growth_real_runtime_ns());
-        total_run_time += growth_times.back();
-        wg->update_last_real_runtime_ns(wg->real_runtime_ns());
-    }
-
-    int i = 0;
-    for (auto& wg : _ready_wgs) {
-        double cpu_actual_use_ratio = 0.0;
-        if (total_run_time != 0) {
-            cpu_actual_use_ratio = ((double)growth_times[i] / (total_run_time));
-        }
-        wg->set_cpu_actual_use_ratio(cpu_actual_use_ratio);
-        i++;
-    }
-}
-
-StatusOr<ScanTask> ScanTaskQueueWithWorkGroup::take(int worker_id) {
+StatusOr<ScanTask> WorkGroupScanTaskQueue::take() {
     std::unique_lock<std::mutex> lock(_global_mutex);
 
-    if (_is_closed) {
-        return Status::Cancelled("Shutdown");
-    }
-    while (_ready_wgs.empty()) {
-        _cv.wait(lock);
+    workgroup::WorkGroupScanSchedEntity* wg_entity = nullptr;
+    while (wg_entity == nullptr) {
         if (_is_closed) {
             return Status::Cancelled("Shutdown");
         }
-    }
 
-    _maybe_adjust_weight();
+        _update_bandwidth_control_period();
+
+        if (_ready_wg_entities.empty()) {
+            _cv.wait(lock);
+        } else if (wg_entity = _take_next_wg(); wg_entity == nullptr) {
+            int64_t cur_ns = MonotonicNanos();
+            int64_t sleep_ns = _bandwidth_control_period_end_ns - cur_ns;
+            if (sleep_ns <= 0) {
+                continue;
+            }
+
+            // All the ready tasks are throttled, so wait until the new period or a new task comes.
+            _cv.wait_for(lock, std::chrono::nanoseconds(sleep_ns));
+        }
+    }
 
-    WorkGroupPtr wg = _select_next_wg(worker_id);
-    if (wg->scan_task_queue()->size() == 1) {
-        _ready_wgs.erase(wg);
+    // If wg only contains one ready driver, it will be not ready anymore
+    // after taking away the only one task.
+    if (wg_entity->queue()->size() == 1) {
+        _sum_cpu_limit -= wg_entity->cpu_limit();
+        _ready_wg_entities.erase(wg_entity);
```

Comment:
Can we use put this code into a function and give a significant name?  in my first sight, I always try to find out in which scenarios that a wg is blocked (I think the opposition 'ready' is 'blocked' in schedule algorithms), but the 'ready' here really means that the private queue of a wg is non-empty and it has drivers ready to run.  for an example:
1. detach_non_ready_wg_from_ready_queue;// for remove wg has to driver in its queue.
2. attach_ready_wg_to_ready_queue; // for add wg has driver to schedule into the queue